src/completion.rs

```diff
@@ -0,0 +1,163 @@
+    }
+}
+
+impl Drop for Completion<'_> {
+    fn drop(&mut self) {
+        // Ensure that after dropping the Completion, the AIO callback
+        // will not be called on our dropped waker Box.  Only necessary
+        // if we got as far as successfully starting an operation using
+        // the completion.
+        let am_complete = unsafe { rados_aio_is_complete(self.inner) } != 0;
+        if !am_complete {
+            unsafe {
+                let cancel_r = rados_aio_cancel(self.ioctx.ioctx, self.inner);
+
+                // It is unsound to proceed if the Objecter op is still in flight
+                assert!(cancel_r == 0 || cancel_r == -libc::ENOENT);
+            }
+        }
+
+        unsafe {
+            // Even if is_complete was true, librados might not be done with
+            // our callback: wait til it is.
+            assert_eq!(rados_aio_wait_for_complete_and_cb(self.inner), 0);
+        }
+
+        unsafe {
+            rados_aio_release(self.inner);
+        }
+    }
+}
+
+impl std::future::Future for Completion<'_> {
+    type Output = crate::error::RadosResult<i32>;
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        // Hold lock across the check of am_complete and subsequent waker registration
+        // to avoid deadlock if callback is invoked in between.
+        let mut waker_locked = self.waker.lock().unwrap();
+
+        let am_complete = unsafe { rados_aio_is_complete(self.inner) } != 0;
+
+        if am_complete {
+            // Unlock Waker so that completion callback can complete if racing with us.
+            drop(waker_locked);
+
+            // Ensure librados is finished with our callback ('complete' is true
+            // before it calls that)
+            unsafe {
+                let r = rados_aio_wait_for_complete_and_cb(self.inner);
+                assert_eq!(r, 0);
+            }
+
+            let r = unsafe { rados_aio_get_return_value(self.inner) };
+            let result = if r < 0 { Err(r.into()) } else { Ok(r) };
+
+            std::task::Poll::Ready(result)
+        } else {
+            // Register a waker
+            *waker_locked = Some(cx.waker().clone());
+
+            std::task::Poll::Pending
+        }
+    }
+}
+
+/// Completions are only created via this wrapper, in order to ensure
+/// that the Completion struct is only constructed around 'armed' rados_completion_t
+/// instances (i.e. those that have been used to start an I/O).
+pub(crate) fn with_completion<F>(ioctx: &IoCtx, f: F) -> RadosResult<Completion<'_>>
+where
+    F: FnOnce(rados_completion_t) -> libc::c_int,
+{
+    let mut waker = Box::new(Mutex::new(None));
+
+    let completion = unsafe {
+        let mut completion: rados_completion_t = std::ptr::null_mut();
+        let p: *mut Mutex<Option<Waker>> = &mut *waker;
+        let p = p as *mut c_void;
+
+        let r = rados_aio_create_completion2(p, Some(completion_complete), &mut completion);
```

Comment:
Please consider this small change, which will ensure the code works with Luminous (current Ceph version in Debian stable):

https://github.com/ankitects/ceph-rust/commit/5966d50f8827d36908bfc93feed4750d5e42a3b5

Reply:
Hmm, luminous is EOL from upstream Ceph's point of view (https://docs.ceph.com/en/latest/releases/index.html).  I think it's appropriate to use the current interface here.